Include JavaFX jmods (and remove forceMerge javaFX)

koppor opened the pull request:
Fixes https://github.com/JabRef/jabref-issue-melting-pot/issues/412

After this is done:

- [x] Report small howto to https://github.com/openjfx/javafx-gradle-plugin/issues/60

### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

koppor commented:
```
Messages are not initialized before accessing key: Display help on command line options
2024-04-09 09:37:26 [JavaFX-Launcher] sun.util.logging.internal.LoggingProviderImpl$JULWrapper.log()
WARN: Unsupported JavaFX configuration: classes were loaded from 'module org.jabref.merged.module', isAutomatic: false, isOpen: true
Exception in thread "JavaFX-Launcher" java.lang.NoSuchMethodError: notifyThemeChanged
        at org.jabref.merged.module@5.14.33/com.sun.glass.ui.win.WinApplication.initIDs(Native Method)
```

Siedlerchr commented:
In 21u it's still present https://github.com/openjdk/jfx21u/blob/c2600685dbba9f33e01b9b0fcf2346df9b5e0011/modules/javafx.graphics/src/main/java/com/sun/glass/ui/Application.java#L262 
In 22 it was changed to a new method: https://github.com/openjdk/jfx/commit/170a4c1545696b095960c0e5dcbed1374d3f6ba6

koppor commented:
Foojay Gradle Plugin seems to stick to Adoptium:

    # Java VM: OpenJDK 64-Bit Server VM Temurin-22+36 (22+36, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, windows-amd64)

koppor commented:
Also happens when using Zulu:

    # JRE version: OpenJDK Runtime Environment Zulu22.28+93-CRaC-CA (22.0+36) (build 22+36)

koppor commented:
Linux:

```
Loading library prism_es2 from resource failed: java.lang.UnsatisfiedLinkError: /home/koppor/.openjfx/cache/23-ea+12/amd64/libprism_es2.so: libXxf86vm.so.1: cannot open shared object file: No such file or directory
java.lang.UnsatisfiedLinkError: /home/koppor/.openjfx/cache/23-ea+12/amd64/libprism_es2.so: libXxf86vm.so.1: cannot open shared object file: No such file or directory
        at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
```

Siedlerchr commented on .github/workflows/deployment-jdk-ea.yml:
```diff
@@ -111,46 +148,30 @@ jobs:
-          echo "Downloading $URL..."
-          curl -OJ --no-progress-meter $URL
+
+          URL_SDK=$(grep -o "https://download.java.net/java/.*/javafx.*${OS}-x64_bin-sdk.${EXT}" javafx.html | head -n 1)
```
duplicate dot

koppor replied:
no, because `EXT="zip"`

Siedlerchr commented:
Oh finally after 10 hours?

koppor commented:
Error: Modules org.jabref.merged.module and javafx.base export package javafx.beans.property to module com.dlsc.gemsfx

Siedlerchr commented on build.gradle:
```diff
@@ -585,7 +585,6 @@ jlink {
     // but it was removed in the newer releases.
     // The pom.xml associated with such a non-modular artifact does not mention that the artifact depends on the removed code
     // (because the artifact was published when this code was still available in the JDK).
-    forceMerge "javafx"
```
this will not work this is intended for our merged module.

Siedlerchr replied:
you could try to add gemsfx to the merged module declaration

koppor replied:
Based on the comments above, I was thinking, it could help.

    Two versions of module javafx.graphics found in /Users/runner/work/***/***/build/jlinkbase/jlinkjars (javafx-graphics-20-mac.jar and javafx.graphics.jar)

Siedlerchr replied:
This sounds like the zulu already ships with javafx and now it cannot decide which one to use because we get javafx over the plugin

koppor replied:
/home/runner/.gradle/caches/modules-2/files-2.1/org.openjfx/javafx-base/20/d82ec648e48647e10125ff0e3dce70348a80ac18/javafx-base-20-linux.jar

Siedlerchr replied:
This is likely the main culprit: 
https://github.com/JabRef/jabref/blob/b140d8a81ba63c439e4502c70c286fcaa590ca2d/build.gradle#L104-L107

koppor replied:
(I document more for myself than for others)

> This is likely the main culprit:

No. 

This points to JavaFX 22, not 22. The file included is ` /home/runner/.gradle/caches/modules-2/files-2.1/org.openjfx/javafx-base/20/d82ec648e48647e10125ff0e3dce70348a80ac18/javafx-base-20-linux.jar` and thus is 20 (twenty) not 22 (twenty-two)

I replace `version` by `sdk` in the `yaml` workflow: https://github.com/JabRef/jabref/blob/f6a7217983d82ea236d6ab3b18643643682e81bb/.github/workflows/deployment-jdk-ea.yml#L166

Siedlerchr replied:
I seee,  but they are still duplicated on the classpath

koppor replied:
Only with the `-20` or do you see other duplicates?

koppor replied:
Maybe https://github.com/beryx/badass-jlink-plugin/issues/61#issuecomment-504640018 helps. Currently checking.

koppor commented:
Adoptium did not work with `22`, therefore switching back to Liberica:

```
  Trying to resolve the latest version from remote
  Error: Could not find satisfied version for SemVer '22'. 
```

koppor commented:
I added the debug too late, nevertheless here the intermediate state:

![image](https://github.com/JabRef/jabref/assets/1366654/7b28e5e7-cce5-43cf-81a1-1f10dfa793f3)

koppor commented:
Error: Modules org.jabref.merged.module and javafx.graphics export package javafx.animation to module jpro.utils.treeshowing

```
+--- com.dlsc.gemsfx:gemsfx:2.7.4
|    \--- one.jpro.jproutils:tree-showing:0.2.2
```

koppor commented:
Local patch (which is applied by the .yaml workflow)

```diff
diff --git a/build.gradle b/build.gradle
index 968bea38d8..26f87bfd49 100644
--- a/build.gradle
+++ b/build.gradle
@@ -103,7 +103,7 @@ dependencyLocking {
 }

 javafx {
-    version = "22"
+    sdk = "javafx/javafx-sdk-23"
     modules = [ 'javafx.controls', 'javafx.fxml', 'javafx.web', 'javafx.swing' ]
 }

@@ -577,6 +577,7 @@ tasks.register('deleteInstallerTemp', Delete) {
 jpackage.dependsOn deleteInstallerTemp
 jlinkZip.dependsOn jpackage
 jlink {
+    addExtraModulePath 'javafx/javafx-jmods-23'
     addOptions('--strip-debug', '--compress', 'zip-6', '--no-header-files', '--no-man-pages')
     launcher {
         name = 'JabRef'
```

koppor commented:
Both the sdk and the jmods are available at https://jdk.java.net/javafx23/. Unzip in `javafx` sub directory.

koppor commented:
- JDK: Corretto (Temurin did not work(
- removed `forceMerge` ([`0c87bdd` (#11170)](https://github.com/JabRef/jabref/pull/11170/commits/0c87bdd1bccfd1c02e330e809d806d381148bb98) and added `requires` statements ([`a96d0f9` (#11170)](https://github.com/JabRef/jabref/pull/11170/commits/a96d0f9c2f77483a4c4f887a8b8ba56aeec5b191)

Result:

![image](https://github.com/JabRef/jabref/assets/1366654/719d5c03-2f96-4fe3-a342-df4fdee10cdb)

Siedlerchr commented:
pipeline still failing? /home/runner/work/jabref/jabref/build/jlinkbase/tmpjars/org.jabref.merged.module/module-info.java:1113: error: module not found: javafx.base

koppor commented:
No clean: BELLSOFT same error, Corretto: no error

Now, trying clean locally.

koppor commented:
With Amazon Corretto:

![image](https://github.com/JabRef/jabref/assets/1366654/7a60620a-4a99-4740-8ea3-6dcbb14c96f5)
